File: motion_tracking_30fps_11_11_25.cpp
// C++ translation of motion_tracking_30fps_11_11_25.py
// Features:
// - Fisheye-safe undistort (falls back on error)
// - Save G-channel RAW frames to raw_capture
// - Farneback optical flow on downscaled image to compute avg_speed
// - Dynamic sensitivity: threshold and min contour area depend on avg_speed
// - BackgroundSubtractorMOG2 tuned for sensitivity (history=200, varThreshold=16, no shadows)
// - Simple per-frame contour-index trails with fading and prediction
// - UDP JSON output: id,x,y,w,h,angle_h,angle_v

#include <opencv2/opencv.hpp>
#include <opencv2/video/background_segm.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/calib3d.hpp>
#include <opencv2/core.hpp>

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <deque>
#include <iomanip>
#include <unordered_map>
#include <filesystem>
#include <cmath>
#include <cstring>

#ifdef _WIN32
  #define _WINSOCK_DEPRECATED_NO_WARNINGS
  #include <winsock2.h>
  #include <ws2tcpip.h>
  #pragma comment(lib, "Ws2_32.lib")
#else
  #include <sys/socket.h>
  #include <arpa/inet.h>
  #include <unistd.h>
#endif

using namespace cv;
using namespace std;

// ---------------- UDP parameters ----------------
static const char* UDP_IP = "192.168.1.100"; // PC IP
static const int UDP_PORT = 5005;

struct UdpSender {
#ifdef _WIN32
    SOCKET sock = INVALID_SOCKET;
#else
    int sock = -1;
#endif
    sockaddr_in addr{};
    bool ok = false;

    UdpSender() {
#ifdef _WIN32
        WSADATA wsaData;
        if (WSAStartup(MAKEWORD(2,2), &wsaData) != 0) {
            cerr << "WSAStartup failed\n";
            return;
        }
        sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
        if (sock == INVALID_SOCKET) {
            cerr << "UDP socket creation failed\n";
            return;
        }
#else
        sock = socket(AF_INET, SOCK_DGRAM, 0);
        if (sock < 0) {
            cerr << "UDP socket creation failed\n";
            return;
        }
#endif
        memset(&addr, 0, sizeof(addr));
        addr.sin_family = AF_INET;
        addr.sin_port = htons(UDP_PORT);
        addr.sin_addr.s_addr = inet_addr(UDP_IP);
        ok = true;
    }

    ~UdpSender() {
#ifdef _WIN32
        if (sock != INVALID_SOCKET) closesocket(sock);
        WSACleanup();
#else
        if (sock >= 0) close(sock);
#endif
    }

    bool sendStr(const std::string& s) {
        if (!ok) return false;
        int r = sendto(sock, s.c_str(), (int)s.size(), 0, (sockaddr*)&addr, (socklen_t)sizeof(addr));
        return r >= 0;
    }
};

// ---------------- Camera settings ----------------
static const int CAM_INDEX = 0;
static const int FPS = 30;
static const int FRAME_WIDTH = 9152;  // match Python
static const int FRAME_HEIGHT = 6944; // match Python

// ---------------- Trail settings ----------------
static const int TRAIL_LENGTH = 20; // number of points to keep in trail
static const int TRAIL_FADE = 10;   // fade step

// ---------------- RAW folder ----------------
static const string RAW_DIR = "raw_capture";

static bool ensure_dir(const string& path) {
    try {
        std::filesystem::create_directories(path);
        return true;
    } catch (const std::exception& e) {
        cerr << "Failed to create directory '" << path << "': " << e.what() << "\n";
        return false;
    }
}

static bool save_raw_g_channel(const Mat& frame, const string& dir, int frame_id) {
    vector<Mat> ch;
    split(frame, ch);
    if (ch.size() < 2) return false;
    const Mat& g = ch[1]; // CV_8UC1
    ostringstream oss;
    oss << dir << "/frame_" << setw(6) << setfill('0') << frame_id << ".raw";
    string path = oss.str();
    ofstream ofs(path, ios::binary);
    if (!ofs) return false;
    ofs.write(reinterpret_cast<const char*>(g.data), (std::streamsize)(g.total() * g.elemSize()));
    return ofs.good();
}

// ---------------- Fisheye-safe undistort ----------------
static Mat undistort_fisheye_safe(const Mat& frame) {
    try {
        int h = frame.rows, w = frame.cols;
        Mat K = (Mat_<float>(3,3) << w/2.0f, 0, w/2.0f,
                                     0,      w/2.0f, h/2.0f,
                                     0,      0,      1);
        Mat D = Mat::zeros(4, 1, CV_32F);
        Mat R = Mat::eye(3, 3, CV_32F);
        Mat map1, map2, out;
        fisheye::initUndistortRectifyMap(K, D, R, K, Size(w, h), CV_16SC2, map1, map2);
        remap(frame, out, map1, map2, INTER_LINEAR, BORDER_CONSTANT);
        return out;
    } catch (...) {
        return frame.clone();
    }
}

// ---------------- Pixel to angle ----------------
static pair<double,double> pixel_to_angle(double x, double y, double width, double height, double fov_horizontal=180.0, double fov_vertical=180.0) {
    double cx = width / 2.0;
    double cy = height / 2.0;
    double nx = (x - cx) / cx;
    double ny = (y - cy) / cy;
    double angle_x = nx * (fov_horizontal / 2.0);
    double angle_y = ny * (fov_vertical / 2.0);
    return {angle_x, angle_y};
}

int main() {
    ensure_dir(RAW_DIR);

    VideoCapture cap(CAM_INDEX);
    cap.set(CAP_PROP_FRAME_WIDTH, FRAME_WIDTH);
    cap.set(CAP_PROP_FRAME_HEIGHT, FRAME_HEIGHT);
    cap.set(CAP_PROP_FPS, FPS);
    if (!cap.isOpened()) {
        cerr << "Failed to open camera index " << CAM_INDEX << "\n";
        return 1;
    }

    // Background subtractor tuned for sensitivity
    Ptr<BackgroundSubtractorMOG2> fgbg = createBackgroundSubtractorMOG2(200, 16, false);

    int frame_id = 0;
    Mat prev_gray_for_flow; // G-channel of previous corrected frame
    int flow_scale = 2;     // downscale factor

    UdpSender udp;

    Mat frame;
    while (true) {
        if (!cap.read(frame)) break;
        frame_id++;

        // Save RAW G-channel
        save_raw_g_channel(frame, RAW_DIR, frame_id);

        // Fisheye correction (safe)
        Mat corrected = undistort_fisheye_safe(frame);

        // Take G channel for processing
        vector<Mat> ch;
        split(corrected, ch);
        Mat gray = (ch.size() >= 2) ? ch[1] : corrected;

        // Farneback optical flow on downscaled images to compute avg_speed
        double avg_speed = 0.0; // average magnitude
        try {
            if (!prev_gray_for_flow.empty()) {
                Size smallSize(gray.cols / flow_scale, gray.rows / flow_scale);
                if (smallSize.width > 0 && smallSize.height > 0) {
                    Mat small_prev, small_curr;
                    resize(prev_gray_for_flow, small_prev, smallSize, 0, 0, INTER_AREA);
                    resize(gray,              small_curr, smallSize, 0, 0, INTER_AREA);
                    Mat flow;
                    calcOpticalFlowFarneback(small_prev, small_curr, flow, 0.5, 3, 15, 3, 5, 1.2, 0);
                    Mat flowXY[2];
                    split(flow, flowXY);
                    Mat mag, ang;
                    cartToPolar(flowXY[0], flowXY[1], mag, ang);
                    Scalar meanMag = mean(mag);
                    avg_speed = meanMag[0];
                }
            }
            prev_gray_for_flow = gray.clone();
        } catch (...) {
            avg_speed = 0.0;
        }

        // Dynamic sensitivity
        int dynamic_thresh = (int)std::lround(std::clamp(50.0 - avg_speed * 10.0, 10.0, 100.0));
        int dynamic_min_area = (int)std::lround(std::clamp(800.0 - avg_speed * 50.0, 200.0, 5000.0));

        // Foreground / motion
        Mat fgmask;
        fgbg->apply(gray, fgmask);
        Mat thresh;
        threshold(fgmask, thresh, dynamic_thresh, 255, THRESH_BINARY);

        vector<vector<Point>> contours;
        findContours(thresh, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);

        struct Obj { int id; int x; int y; int w; int h; double angle_h; double angle_v; int cx; int cy; };
        vector<Obj> objects;

        static unordered_map<int, deque<Point>> trails; // per-frame index based

        for (int i = 0; i < (int)contours.size(); ++i) {
            const auto& cnt = contours[i];
            double area = contourArea(cnt);
            if (area < dynamic_min_area) continue;

            Rect r = boundingRect(cnt);
            int cx = r.x + r.width / 2;
            int cy = r.y + r.height / 2;
            auto ang = pixel_to_angle((double)cx, (double)cy, (double)corrected.cols, (double)corrected.rows);

            objects.push_back(Obj{i, r.x, r.y, r.width, r.height, ang.first, ang.second, cx, cy});

            if (!trails.count(i)) trails[i] = deque<Point>();
            auto& dq = trails[i];
            dq.push_back(Point(cx, cy));
            while ((int)dq.size() > TRAIL_LENGTH) dq.pop_front();

            // Draw rectangle and center
            rectangle(corrected, r, Scalar(0, 255, 0), 2);
            circle(corrected, Point(cx, cy), 4, Scalar(0, 0, 255), FILLED);

            // Draw trail with fading
            for (size_t j = 1; j < dq.size(); ++j) {
                int alpha = max(0, 255 - (int)(dq.size() - j) * TRAIL_FADE);
                line(corrected, dq[j-1], dq[j], Scalar(255, alpha, alpha), 2);
            }

            // Prediction from last motion vector
            if (dq.size() >= 2) {
                Point p1 = dq[dq.size()-1];
                Point p0 = dq[dq.size()-2];
                Point pred(p1.x + (p1.x - p0.x) * 5, p1.y + (p1.y - p0.y) * 5);
                circle(corrected, pred, 6, Scalar(255, 255, 0), 2);
            }
        }

        // UDP send JSON if objects exist
        if (!objects.empty()) {
            ostringstream oss;
            oss.setf(std::ios::fixed);
            oss << "[";
            for (size_t i = 0; i < objects.size(); ++i) {
                const auto& o = objects[i];
                if (i) oss << ",";
                oss << "{"
                    << "\"id\":" << o.id << ","
                    << "\"x\":" << o.x << ","
                    << "\"y\":" << o.y << ","
                    << "\"w\":" << o.w << ","
                    << "\"h\":" << o.h << ","
                    << "\"angle_h\":" << setprecision(6) << o.angle_h << ","
                    << "\"angle_v\":" << setprecision(6) << o.angle_v
                    << "}";
            }
            oss << "]";
            string payload = oss.str();
            udp.sendStr(payload);
            try {
                cout << "[Frame " << frame_id << "] Sent " << objects.size() << " objects to " << UDP_IP << ":" << UDP_PORT << "\n";
                cout << payload << "\n";
            } catch (...) {
                cerr << "Failed to print payload\n";
            }
        }

        imshow("Motion Tracking Fisheye", corrected);
        int key = waitKey(1) & 0xFF;
        if (key == 'q') break;
    }

    cap.release();
    destroyAllWindows();
    return 0;
}

File: motion_tracking_30fps_11_11_25.py
#!/usr/bin/python3
#motion_tracking_30fps_11_11_25.py
#========================
# находит движущиеся объекты
#==============================
# Изменения в этой версии (кратко):
# - Добавлена обработка fisheye с защитой: попытка undistort через cv2.fisheye,
#   при ошибке возвращается исходный кадр, чтобы цикл не падал.
# - Сохранение G-канала каждого кадра в RAW-файлы в папке `raw_capture`.
# - Добавлен быстрый подсчёт средней «скорости» движения по кадрам через
#   Farneback optical flow (уменьшенное изображение) — используется как
#   метрика для адаптивной настройки чувствительности.
# - Динамическая чувствительность вычитания фона: порог бинаризации и
#   минимальная площадь контура зависят от средней скорости движения.
# - Настройки BackgroundSubtractorMOG2 скорректированы для большей чувствительности
#   (меньше history, меньше varThreshold, detectShadows=False).
# - Реализован «след» объектов (trail) с ограниченной длиной и визуальным
#   затуханием (TRAIL_LENGTH, TRAIL_FADE).
# - Рисуется прогнозная точка по вектору последнего движения для каждого следа.
# - Данные об обнаруженных объектах отправляются по UDP в JSON-формате и
#   кратко логируются в консоль для отладки.
#
# Примечания/следующие улучшения (не реализовано здесь):
# - Калибровка метрики "метры/пиксель" (можно добавить диалог ввода и
#   сохранение параметров); для точных скоростей нужна полноценная
#   калибровка камеры.
# - Более устойчивый трекер (SORT/DeepSORT) вместо простого индексирования
#   по контурной области.
#
# - Добавлена функция sharpen_image для повышения резкости изображения
#   с помощью фильтра unsharp masking. Силу резкости можно регулировать
#   параметром --sharpen.
#
import cv2
import numpy as np
import socket
import json
import os
from collections import deque
import argparse
import sys
# Try to import Picamera2 (Raspberry Pi). If unavailable, fall back to OpenCV VideoCapture.
picamera2_available = False
try:
    sys.path.insert(0, '/usr/lib/python3/dist-packages')
    from picamera2 import Picamera2
    picamera2_available = True
except Exception:
    Picamera2 = None
    picamera2_available = False

# Parse command line arguments
parser = argparse.ArgumentParser(description='Motion tracking with fisheye correction and UDP output.')
parser.add_argument('--width', type=int, default=2400, help='Frame width')
parser.add_argument('--height', type=int, default=1200, help='Frame height')
parser.add_argument('--udp-ip', default="192.168.1.100", help='UDP IP address')
parser.add_argument('--udp-port', type=int, default=5005, help='UDP port')
parser.add_argument('--sharpen', type=float, default=0.0, help='Sharpening strength (0.0 to disable, typical 1.5-2.5)')
parser.add_argument('--source', default='0', help='Video source: camera index or path to video file (default 0).')
parser.add_argument('--use-picam2', action='store_true', default=picamera2_available, help='Prefer Picamera2 if available (default auto-detected)')
parser.add_argument('--demo-image', default=None, help='Path to an image file to use as a static demo input')
args = parser.parse_args()

# --- Sharpening Function ---
def sharpen_image(image, strength=1.5):
    """Apply unsharp mask sharpening.

    strength: strength of the effect (0 -> no change). Typical range 0.0-3.0.
    """
    if strength <= 0.0:
        return image
    # Gaussian blur with (0,0) kernel lets OpenCV pick size from sigma
    blurred = cv2.GaussianBlur(image, (0, 0), sigmaX=strength, sigmaY=strength)
    sharpened = cv2.addWeighted(image, 1.0 + strength, blurred, -strength, 0)
    return np.clip(sharpened, 0, 255).astype(np.uint8)

# ---------------- UDP параметры ----------------
UDP_IP = args.udp_ip   # IP ПК
UDP_PORT = args.udp_port
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
 
# ---------------- Настройки камеры ----------------
CAM_INDEX = 2
FPS = 30
FRAME_WIDTH = args.width
FRAME_HEIGHT = args.height
 
# ---------------- Настройки trail ----------------
TRAIL_LENGTH = 20   # сколько точек хранить для "следа"
TRAIL_FADE = 10    # шаг затухания (чем больше — быстрее исчезает)
 
# ---------------- Папка для RAW ----------------
RAW_DIR = "raw_capture"
os.makedirs(RAW_DIR, exist_ok=True)
 
# Очереди для хранения траекторий объектов
trails = {}
 
# ---------------- Функция: коррекция fisheye ----------------
def undistort_fisheye(frame):
    h, w = frame.shape[:2]
    # Матрица камеры (примерная, требует калибровки для точности)
    K = np.array([[w/2, 0, w/2],
                  [0, w/2, h/2],
                  [0,   0,   1]], dtype=np.float32)
    D = np.zeros(4)
    # если нет калибровочных данных, оставляем 0
    try:
        map1, map2 = cv2.fisheye.initUndistortRectifyMap(
            K, D, np.eye(3, dtype=np.float32), K, (w, h), cv2.CV_16SC2
        )
        return cv2.remap(frame, map1, map2, interpolation=cv2.INTER_LINEAR, borderMode=cv2.BORDER_CONSTANT)
    except Exception:
        # при ошибке возвращаем оригинал, чтобы цикл не упал
        return frame


# ---------------- Перевод пикселей в углы ----------------
def pixel_to_angle(x, y, width, height, fov_horizontal=180, fov_vertical=180):
    cx = width / 2
    cy = height / 2
    nx = (x - cx) / cx   # нормализуем [-1;1]
    ny = (y - cy) / cy
    angle_x = nx * (fov_horizontal / 2)
    angle_y = ny * (fov_vertical / 2)
    return angle_x, angle_y

# ---------------- Capture initialization ----------------
cap = None
use_picam2 = False
picam2 = None

def _init_picam2():
    p = Picamera2()
    cfg = p.create_video_configuration(main={"size": (FRAME_WIDTH, FRAME_HEIGHT)}, controls={"FrameRate": FPS})
    p.configure(cfg)
    p.start()
    return p

if args.demo_image:
    demo_img = cv2.imread(args.demo_image, cv2.IMREAD_COLOR)
    if demo_img is None:
        print(f"Failed to load demo image: {args.demo_image}")
        sys.exit(1)
    demo_img = cv2.resize(demo_img, (FRAME_WIDTH, FRAME_HEIGHT))
    def get_frame():
        return demo_img.copy()
else:
    if args.use_picam2 and picamera2_available:
        try:
            picam2 = _init_picam2()
            use_picam2 = True
            def get_frame():
                f = picam2.capture_array()
                return cv2.cvtColor(f, cv2.COLOR_RGB2BGR)
        except Exception as e:
            print(f"Failed to initialize Picamera2: {e}. Falling back to OpenCV VideoCapture.")
            use_picam2 = False

    if not use_picam2:
        # Interpret numeric source as camera index
        src = args.source
        try:
            idx = int(src)
            cap = cv2.VideoCapture(idx)
        except Exception:
            cap = cv2.VideoCapture(src)

        if not cap or not cap.isOpened():
            print(f"Failed to open video source: {args.source}")
            sys.exit(1)

        # set desired resolution when possible
        try:
            cap.set(cv2.CAP_PROP_FRAME_WIDTH, FRAME_WIDTH)
            cap.set(cv2.CAP_PROP_FRAME_HEIGHT, FRAME_HEIGHT)
            cap.set(cv2.CAP_PROP_FPS, FPS)
        except Exception:
            pass

        def get_frame():
            ret, frame = cap.read()
            if not ret:
                return None
            return frame

# Более чувствительная настройка вычитания фона:
# - уменьшён history чтобы модель быстрее адаптировалась к изменениям
# - уменьшён varThreshold чтобы обнаруживать более слабые изменения
# - отключено обнаружение теней (меньше ложных срабатываний)
fgbg = cv2.createBackgroundSubtractorMOG2(history=200, varThreshold=16, detectShadows=False)

frame_id = 0
prev_gray_for_flow = None
flow_scale = 2  # downscale factor for optical flow to speed up computation

while True:
    frame = get_frame()
    if frame is None:
        # source ended or read error
        print("No frame received (source ended). Exiting.")
        break

    frame_id += 1

    # Сохраняем RAW G-канал
    g_channel = frame[:, :, 1]
    raw_path = os.path.join(RAW_DIR, f"frame_{frame_id:06d}.raw")
    g_channel.tofile(raw_path)

    # Коррекция fisheye
    corrected = undistort_fisheye(frame)
    # optionally apply sharpening
    if args.sharpen > 0:
        corrected = sharpen_image(corrected, strength=args.sharpen)
    #corrected = frame


    # Берём G-канал для обработки
    gray = corrected[:, :, 1]

    # Оценка средней скорости движения между кадрами (Farneback на уменьшенном изображении)
    avg_speed = 0.0
    try:
        if prev_gray_for_flow is not None:
            small_prev = cv2.resize(prev_gray_for_flow, (gray.shape[1]//flow_scale, gray.shape[0]//flow_scale))
            small_curr = cv2.resize(gray, (gray.shape[1]//flow_scale, gray.shape[0]//flow_scale))
            # convert to float32
            small_prev_f = small_prev.astype(np.uint8)
            small_curr_f = small_curr.astype(np.uint8)
            flow = cv2.calcOpticalFlowFarneback(small_prev_f, small_curr_f, None,
                                                0.5, 3, 15, 3, 5, 1.2, 0)
            mag, ang = cv2.cartToPolar(flow[..., 0], flow[..., 1])
            avg_speed = float(np.mean(mag))
        # store current for next frame
        prev_gray_for_flow = gray.copy()
    except Exception:
        avg_speed = 0.0

    # Динамическая настройка чувствительности: чем выше avg_speed, тем ниже порог (более чувствительно)
    # порог бинаризации в диапазоне [10, 100]
    dynamic_thresh = int(np.clip(50 - avg_speed * 10.0, 10, 100))
    # минимальная площадь контура: чем быстрее, тем меньше (ловим маленькие быстрые объекты)
    dynamic_min_area = int(np.clip(800 - avg_speed * 50.0, 200, 5000))

    # Фон/движение
    fgmask = fgbg.apply(gray)
    # Пониженный/динамический порог бинаризации — реагируем на слабые изменения фона
    _, thresh = cv2.threshold(fgmask, dynamic_thresh, 255, cv2.THRESH_BINARY)
   
    contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    objects = []

    for i, cnt in enumerate(contours):
        # динамический порог по площади контура — ловим меньше объектов при медленном движении
        if cv2.contourArea(cnt) < dynamic_min_area: # фильтр шума
            continue

        x, y, w, h = cv2.boundingRect(cnt)
        cx, cy = x + w // 2, y + h // 2

        # вычисляем углы
        angle_h, angle_v = pixel_to_angle(cx, cy, corrected.shape[1], corrected.shape[0])

        obj = {
            "id": i,
            "x": int(x),
            "y": int(y),
            "w": int(w),
            "h": int(h),
            "angle_h": float(angle_h),
            "angle_v": float(angle_v)
        }
        objects.append(obj)

        # --- Trail обработка ---
        if i not in trails:
            trails[i] = deque(maxlen=TRAIL_LENGTH)
        trails[i].append((cx, cy))

        # Рисуем рамку и центр
        cv2.rectangle(corrected, (x, y), (x+w, y+h), (0, 255,0), 2)
        cv2.circle(corrected, (cx, cy), 4, (0, 0, 255), -1)

        # Рисуем след с затуханием
        pts = list(trails[i])
        for j in range(1, len(pts)):
            alpha = max(0, 255 - (len(pts)-j)*TRAIL_FADE)
            cv2.line(corrected, pts[j-1], pts[j], (255, alpha, alpha), 2)

        # Прогноз: продолжаем движение по последнему вектору
        if len(pts) >= 2:
            dx = pts[-1][0] - pts[-2][0]
            dy = pts[-1][1] - pts[-2][1]
            pred_x, pred_y = pts[-1][0] + dx*5, pts[-1][1] + dy*5
            cv2.circle(corrected, (pred_x, pred_y), 6, (255, 255, 0), 2)

    # --- Отправка данных на ПК ---
    if objects:
        data = json.dumps(objects).encode("utf-8")
        sock.sendto(data, (UDP_IP, UDP_PORT))
        # Выводим отправляемые данные в терминал для отладки (кратко)
        try:
            print(f"[Frame {frame_id}] Sent {len(objects)} objects to {UDP_IP}:{UDP_PORT}")
            print(json.dumps(objects, ensure_ascii=False))
        except Exception as e:
            print(f"Failed to print payload: {e}")

    # --- Показ ---
    cv2.imshow("Motion Tracking Fisheye", corrected)
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

picam2.stop()
cv2.destroyAllWindows()
if cap is not None:
    try:
       cap.release()
    except Exception:
       pass

#if 'picam2' in globals() and use_picam2:
#    try:
#        picam2.stop()
#   except Exception:
#        pass

#[DEBUG ON]
#[DEBUG OFF]

#================================ RESTART: Shell ================================

